c2rust-analyze/src/analyze.rs

```diff
@@ -993,7 +993,10 @@ fn run(tcx: TyCtxt) {
                 if !node_info.unique {
                     perms.remove(PermissionSet::UNIQUE);
                 }
-                // TODO: PermissionSet::NON_NULL
+                if g.is_null {
+                    // TODO: is this enough?
+                    perms.remove(PermissionSet::NON_NULL);
+                }
 
```

Comment:
@spernsteiner If I'm understanding #1088 correctly, the other thing we need to do here is add `NON_NULL` to `updates_forbidden`?

Reply:
Roughly, yes.  Though we might want to experiment a bit - for example, it might be easier to understand the results if we set `updates_forbidden` only on function signatures, or only on named variables and not on temporaries.

Reply:
I think there can be two nodes with the same `dest` (either in different `Graph`s or in the same one), representing two different observations of that `dest` during the execution.  We want to set `NON_NULL` in `perms` (and in `updates_forbidden`) only if all observations with the same `dest` show it containing a non-null pointer.  So we need some additional logic beyond just doing `perms.insert(NON_NULL); updates_forbidden.insert(NON_NULL);` here.